federation/planner.go

```diff
@@ -0,0 +1,368 @@
+
+const (
+	KindType StepKind = iota
+	KindField
+)
+
+// PathStep represents a step in the path from the original query to a subquery that can be
+// resolved on a single GraphQL server. Lets go through a few examples
+//
+// If we have a selection type like the example below
+// previouspathstep {
+//  a: s1nest
+// }
+// The list of path steps should include {Kind: KindField, Name: "previouspathstep"} and {Kind: KindField, Name: "a"} to indicate this
+// subquery is nested on "previouspathstep" and "a"
+//
+// If we have a union type like the example below,
+// previouspathstep {
+// 	... on Foo {
+// 		name
+// 	}
+// }
+// The list of path steps should include  {Kind: KindField, Name: "previouspathstep"} and {Kind: KindType, Name: "Foo"} to indicate this
+// subquery is nested on "previouspathstep" selection and the "Foo" type.
+type PathStep struct {
+	Kind StepKind // StepKind is used to indicate the type of previous steps in the plan are
+	Name string   // Name of the selection or type this path is nested on
+}
+
+// Plan breaks the query down into subqueries that can be resolved by a single graphql server
+type Plan struct {
+	Path         []PathStep            // Pathstep defines what the steps this subplan is nested on
+	Service      string                // Service that resolves this path step
+	Kind         string                // Kind is either a query or mutation
+	Type         string                // Type is the name of the object type each subplan is nested on
+	SelectionSet *graphql.SelectionSet // Selections that will be resolved in this part of the plan
+	After        []*Plan               // Subplans from nested queries on this path
+}
+
+// Planner is responsible for taking a query created a plan that will be used by the executor.
+// This breaks every query into subqueries that can each be resolved by a single graphQLServer
+// and describes what sub-queries need to be resolved first.
+type Planner struct {
+	schema    *SchemaWithFederationInfo //schema describes what fields the graphql servers know about along with the services that know how to execute each field
+	flattener *flattener                //flattener knows how to combine all the fragments on a query into a singel query
+}
+
+// Executing a subquery
+//
+// When a subquery is run on a seperate graphql server, we want the subquery to be nested
+// on the "__federation" type so that the graphQL server
+// For example, if our query like the example below where "allFoos" is a field on service1
+// and "name" is a field on service2.
+// {
+// 	allFoos {
+// 		name
+// 	}
+// }
+// When we send the subquery to service2 it should look like the example below
+// {
+// 	__federation {
+// 		fooById(id: $id) {
+//      __typename
+// 			name
+// 		}
+// 	}
+// }
+// "__federation" becomes the root query that the subquery is nested under,
+// "fooById" is the federated object type that we need to refetch,
+// and "__typename" lets us know what type the object is.
```

Comment:
does "us" mean the gateway or the schema server?

Reply:
the gateway, just changed it!